include/aws/crt/http/HttpConnection.h

```diff
@@ -0,0 +1,356 @@
+                 * to close them before shutting down your application or whatever is consuming this interface.
+                 *
+                 * Assuming `OnConnectionShutdown` has not already been invoked, it will be invoked as a result of this
+                 * call. It is safe to release your reference to this object after calling this function.
+                 */
+                bool Close() noexcept;
+
+                int LastError() const noexcept { return m_lastError; }
+
+              private:
+                HttpConnection(aws_http_connection *m_connection, Allocator *allocator) noexcept;
+                aws_http_connection *m_connection;
+                Allocator *m_allocator;
+                int m_lastError;
+
+                friend class HttpClient;
+            };
+
+            /**
+             * Represents an Http Client. `onConnectionSetup` will be invoked upon an outgoing connection,
+             * and `onConnectionShutdown` will be invoked when a connection is shutdown.
+             */
+            class AWS_CRT_CPP_API HttpClient final
+            {
+              public:
+                /**
+                 * Initializes client with a bootstrap object. `initialWindowSize` will set the TCP read window
+                 * allowed for Http 1.1 connections and Initial Windows for H2 connections.
+                 */
+                HttpClient(
+                    Io::ClientBootstrap *bootstrap,
+                    std::size_t initialWindowSize = SIZE_MAX,
+                    Allocator *allocator = DefaultAllocator()) noexcept;
+
+                /**
+                 * Establish a new Https Connection to hostName:port, using `socketOptions` for tcp options and
+                 * `tlsConnOptions` for TLS/SSL options.
+                 *
+                 * returns true on success, and false on failure. If false is returned, `onConnectionSetup` will not
+                 * be invoked.
+                 */
+                bool NewConnection(
+                    const ByteCursor &hostName,
+                    uint16_t port,
+                    const Io::SocketOptions &socketOptions,
+                    const Io::TlsConnectionOptions &tlsConnOptions) const noexcept;
+
+                /**
+                 * Establish a new Http (Plain-text) Connection to hostName:port, using `socketOptions` for tcp options.
+                 *
+                 * returns true on success, and false on failure. If false is returned, `onConnectionSetup` will not
+                 * be invoked.
+                 */
+                bool NewConnection(const ByteCursor &hostName, uint16_t port, const Io::SocketOptions &socketOptions)
+                    const noexcept;
+
+                /**
+                 * See `OnConnectionSetup` for more info. This value cannot be empty.
+                 */
+                OnConnectionSetup onConnectionSetup;
```

Comment:
For all these values that cannot be empty, should we make them ctor params (even if that ctor is an overload)? I feel like otherwise we're setting the user up for the WINAPI problem of EINVALID_XXX, Lassie-style debugging/appeasement of the API. I hate to suggest a builder, but this use case is something it's good at.